packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx

```diff
@@ -61,13 +61,16 @@ function UnforwardedToggleGroupControlAsRadioGroup(
 		if ( previousValue !== radio.state ) {
 			onChange( radio.state );
 		}
-	}, [ radio.state ] );
+	}, [ radio.state, previousValue, onChange ] );
 
 	// Sync incoming value with radio.state.
 	useUpdateEffect( () => {
 		if ( value !== radio.state ) {
 			radio.setState( value );
 		}
+		// disable reason: Adding `radio` to the dependency array causes an extra render.
```

Comment:
What if we did something like

```js
const setRadioState = radio.setState;

// ...

useUpdateEffect( () => {
		if ( value !== radio.state ) {
			setRadioState( value );
		}
}, [ setRadioState, value ] );
```

That way, the hook will re-run only when `setRadioState` changes (and not the whole `radio` object). Furthermore, there's a chance that `setState` is an actual React `useState` setter, which means its value will be stable throughout the component's lifetime (although we should check for that)

Reply:
Good idea - playing with it got interesting though:

If we go this route, we then need `radio.state` as a dependency. ESLint wasn't asking for that before, perhaps because it would have been happy with the full `radio` object?

Anyway - if we add _that_ dependency, we end up (once again) with an additional render whenever a new option is selected, just like we'd get if we added `radio` directly. I suppose that makes sense, since `radio.state` is going to change every time.

This leaves me uncertain of the best course forward. Leave the disabling comment for now? Just add `radio` as the linter originally wanted, rather than introduce a new variable that doesn't actually solve our issue? 🤔

Reply:
Good point. Adding those dependencies sounds correct to me, though — we could apply the technique that I mentioned above to both `state` and `setState`, and do something like this

<details>

<summary>Click to expand</summary>

```diff
diff --git a/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx b/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx
index 8cf75e2266..a2fc579d2b 100644
--- a/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx
+++ b/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx
@@ -64,14 +64,14 @@ function UnforwardedToggleGroupControlAsRadioGroup(
 	}, [ radio.state, previousValue, onChange ] );
 
 	// Sync incoming value with radio.state.
+	const { state: radioState, setState: setRadioState } = radio;
 	useUpdateEffect( () => {
-		if ( value !== radio.state ) {
-			radio.setState( value );
+		if ( value !== radioState ) {
+			setRadioState( value );
 		}
-		// disable reason: Adding `radio` to the dependency array causes an extra render.
-		// Memoizing it doesn't look easily doable, so we're disabling the rule for now.
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ value ] );
+		// setRadioState needs to be listed even if in theory it's supposed to be a
+		// stable reference — that's an ESLint limitation.
+	}, [ radioState, setRadioState, value ] );
 
 	return (
 		<ToggleGroupControlContext.Provider
```
</details>

We'd need of course to perform some testing and make sure that this doesn't introduce regressions — but, in theory, the hook is relatively simple and it makes sense to me that we re-run the hook when the radio state updates

Reply:
Left a couple additional suggestions. On top of that, I think that it would be best to update the `as-radio-group.tsx` file with the same pattern (using `useControlledState` for handling controlled/uncontrolled updates, creating the context value separately with `useMemo`), even if it's not 100% related to original scope of the PR